### ISSUE TITLE
`--throttle-upload` affects restore download performance

### ISSUE THREAD

**Issue by JonMikelV** [duplicati/duplicati#3272]
- [ x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.3.4 & 2.0.3.6
- **Operating system**: Any
- **Backend**: Any

## Description

`--throttle-upload` (aka "Max upload speed") seems to be getting applied to DOWNLOADS such as during a restore.  This seems to be happening to both remote AND LOCAL DRIVE restores (see https://github.com/duplicati/duplicati/issues/2685).

This issue was prompted by a user report on the Duplicati forms at:
https://forum.duplicati.com/t/throttle-upload-effects-restores/2632

Note that the `--throttle-download` setting seems to have NO EFFECT on performance so I'm guessing this is related to a similar issue I seem to recall where parameter order was incorrect.

## Steps to reproduce

1. Set a job with the following parameters:
  a. `+*DownloadSpeed*;-*`
  a. `--log-file=[path to log file]` (to see reported speeds)
  b. `--log-file-log-level=Profiling` (for download speed details)
  c. `--log-file-log-filter=+*DownloadSpeed*;-*` (to limit log to just what we care about)
  d. `--no-local-blocks=true` (so only backed up blocks are used during restore)
  e. `--throttle-upload=0kb` (NO throttling)
2. Do a restore
4. Set `--throttle-upload=1MB` (or similar)
5. Re-do the restore
6. Set `--throttle-upload=512kb` (or similar)
7. Re-do the restore
8. Look at [path to log file] for "BackendManager-DownloadSpeed" lines such as included below & note how the MB/s times vary based on `--throttle-upload` changes:

- **Actual result**:
BackednManager-DownloadSpeed log entries show MB/s rates approx. the same as `--throttle-upload` setting.
- **Expected result**:
The `--throttle-upload` setting should have NO effect on DOWNLOAD rates such as during a restore.

## Debug log

~~~
2018-06-14 13:57:20 -05 - [Profiling-Duplicati.Library.Main.BackendManager-DownloadSpeed]: Downloaded 3.81 MB in 00:00:00.2730619, 13.96 MB/s
2018-06-14 13:58:02 -05 - [Profiling-Duplicati.Library.Main.BackendManager-DownloadSpeed]: Downloaded 3.81 MB in 00:00:03.2120432, 1.19 MB/s
2018-06-14 13:58:44 -05 - [Profiling-Duplicati.Library.Main.BackendManager-DownloadSpeed]: Downloaded 3.81 MB in 00:00:06.7634094, 576.99 KB/s
~~~

**Comment by chriswm** [comment]
I believe the cause of this bug is a mistake in the order the parameters are passed to the `ThrottledStream` object.

- The `ThrottledStream` object takes a `readspeed` (download) as the second parameter and `writespeed` (upload) as the third parameter. https://github.com/duplicati/duplicati/blob/aa1f99628cddb3630d059d74cb38a9f3424082bf/Duplicati/Library/Utility/ThrottledStream.cs#L108
- The `BackEndHandler` passes the `m_options.MaxUploadPrSecond` as the second parameter and `m_options.MaxDownloadPrSecond` as the third parameter https://github.com/duplicati/duplicati/blob/aa1f99628cddb3630d059d74cb38a9f3424082bf/Duplicati/Library/Main/Operation/Common/BackendHandler.cs#L462

I found this by code inspection (I encountered the same problem when I attempted to limit the upload speed in the Web UI). I don't have a dev environment set-up to test this and submit a fix though but figured I'd record what I found in case I find the time to set-up a dev environment in the future, or if it could make someone else's job fixing this a bit easier.

**Comment by Pectojin** [comment]
Definitely a little funky... 2 references calling download,upload and 3 calling upload,download
<img width="854" alt="screen shot 2018-10-08 at 23 19 32" src="https://user-images.githubusercontent.com/1561484/46634299-ad4b1000-cb50-11e8-853b-f2e485454b34.png">

Some in are "Get" calls, while others are in "Put" calls, but it seems counter intuitive that it wouldn't always be downloading -> reading.

**Comment by ts678** [comment]
Bumped into this while using web UI "Throttle settings" (at top of screen) to set "Max upload speed" to `200 KByte/sec` to see what GUI would show on slow upload (investigating https://forum.duplicati.com/t/upload-speed-to-google-drive/6054 and wondering if reported `200kb/` was actually 200KByte/sec, i.e. units error).

I never did see a metric flash by, but verify step was slow in the GUI, About --> System info --> Server state properties showed it doing a slow dblock GET, Task Manager showed bursty-but-slow receive, and no send:

The view from Profiling log looks like download speed got capped to my upload setting, but upload did not.

~~~
2019-01-19 08:44:21 -05 - [Profiling-Duplicati.Library.Main.Operation.Common.BackendHandler-UploadSpeed]: Uploaded 8.99 MB in 00:00:16.7190026, 550.69 KB/s
2019-01-19 08:44:22 -05 - [Profiling-Duplicati.Library.Main.Operation.Common.BackendHandler-UploadSpeed]: Uploaded 11.08 KB in 00:00:00.2853985, 38.81 KB/s
2019-01-19 08:44:22 -05 - [Profiling-Duplicati.Library.Main.Operation.Common.BackendHandler-UploadSpeed]: Uploaded 30.87 KB in 00:00:00.3124803, 98.80 KB/s
2019-01-19 08:44:25 -05 - [Profiling-Duplicati.Library.Main.BackendManager-UploadSpeed]: Uploaded 8.97 KB in 00:00:02.5932240, 3.46 KB/s
2019-01-19 08:44:27 -05 - [Profiling-Duplicati.Library.Main.BackendManager-DownloadSpeed]: Downloaded 30.87 KB in 00:00:01.0846023, 28.46 KB/s
2019-01-19 08:44:28 -05 - [Profiling-Duplicati.Library.Main.BackendManager-DownloadSpeed]: Downloaded 30.03 KB in 00:00:00.7764771, 38.67 KB/s
2019-01-19 08:48:47 -05 - [Profiling-Duplicati.Library.Main.BackendManager-DownloadSpeed]: Downloaded 50.07 MB in 00:04:18.7814622, 198.12 KB/s
~~~

so I went searching for a throttling issue I recalled. Here it is, and I guess it's still happening in 2.0.4.5 beta.
